src/sql/base/browser/ui/taskbar/actionbar.ts

```diff
@@ -128,15 +143,87 @@ export class ActionBar extends ActionRunner implements IActionRunner {
 
+		if (this._collapseOverflow) {
+			this._overflow = document.createElement('ul');
+			this._overflow.id = 'overflow';
+			this._overflow.className = 'overflow';
+			this._domNode.appendChild(this._overflow);
+		} else {
+			this._actionsList.style.flexWrap = 'wrap';
+		}
+
 		container.appendChild(this._domNode);
 	}
 
+	@debounce(300)
+	private resizeToolbar() {
+		let width = document.getElementById('actions-container').offsetWidth;
+		let fullWidth = document.getElementById('actions-container').scrollWidth;
+
+		// collapse actions that are beyond the width of the toolbar
+		if (width < fullWidth) {
+			// create '•••' more element if it doesn't exist yet
+			if (!this._moreItemElement) {
+				this._moreItemElement = document.createElement('li');
+				this._moreItemElement.className = 'action-item more';
+				this._moreItemElement.setAttribute('role', 'presentation');
+				this._moreItemElement.id = 'more';
+				const innerText: HTMLElement = document.createElement('a');
+				innerText.setAttribute('role', 'button');
+				innerText.innerHTML = '...';
+				innerText.style.fontWeight = 'bold';
+				innerText.style.fontSize = 'large';
+				innerText.title = nls.localize('toggleMore', "Toggle More");
+				innerText.tabIndex = 0;
+				innerText.onclick = (this._domNode, ev => { this._overflow.style.display = this._overflow.style.display === 'block' ? 'none' : 'block'; });
+				innerText.onkeydown = (this._domNode, ev => {
+					let event = new StandardKeyboardEvent(ev);
+					if (event.keyCode === KeyCode.Enter || event.keyCode === KeyCode.Space) {
+						this._focusedItem = undefined; // so that the default actionbar click handler doesn't trigger the selected action-item
+						this._overflow.style.display = this._overflow.style.display === 'block' ? 'none' : 'block';
+						event.preventDefault();
+						event.stopPropagation();
+					}
+				});
+				this._moreItemElement.appendChild(innerText);
+				this._actionsList.appendChild(this._moreItemElement);
+			}
+
+			this._moreItemElement.style.display = 'block';
+			while (width < fullWidth) {
+				let index = this._actionsList.childNodes.length - 2;
+				if (index > -1) {
+					let item = this._actionsList.removeChild(this._actionsList.childNodes[index]);
+					this._overflow.insertBefore(item, this._overflow.firstChild);
+					fullWidth = document.getElementById('actions-container').scrollWidth;
+				} else {
+					break;
+				}
+			}
+		} else if (this._overflow?.hasChildNodes()) { // uncollapse actions if there is space for it
+			while (width === fullWidth && this._overflow.hasChildNodes()) {
+				this._actionsList.insertBefore(this._overflow.removeChild(this._overflow.firstChild), this._actionsList.lastChild);
+
+				// if the action was too wide, collapse it again
+				if (document.getElementById('actions-container').scrollWidth > document.getElementById('actions-container').offsetWidth) {
+					let index = this._actionsList.childNodes.length - 2;
```

Comment:
Some duplicated logic here - would be good to see that moved into a helper method